### PR TITLE
feat(@aws-amplify/auth): Allow bypassing cache on setPreferredMFA and getPreferredMFA methods

### LINKED ISSUE #2627: Cannot set MFA to `NOMFA` immediately after setting it to `SMS` or `TOTP`
**Describe the bug**
If I try to set MFA type to `NOMFA` **immediately** after setting it to `SMS` or `TOTP` via `Auth.setPreferredMFA`, the preferred MFA method of the user doesn't change.

**To Reproduce**
1. Sign in with a user who hasn't configured MFA
2. **Immediately** set his MFA method to `SMS` or `TOTP` via `Auth.setPreferredMFA(user, "SMS")`
3. Set his MFA method to `NOMFA` via `Auth.setPreferredMFA(user, "NOMFA")`
4. MFA method is always set to `SMS` (or `TOTP`)

And I get the following error: `No change for mfa type`

**Expected behavior**
The MFA type should be set to `NOMFA`.

**Desktop (please complete the following information):**
 - OS: macOS Sierra
 - Browser: Chrome
 - Version: 71.0.3578.98

## Notes
After some research and debug, I noticed that the bug doesn't occur for a user that have already set hist MFA method to `SMS` or `TOTP` (the bug also doesn't occur anymore after sign out/sign in).

I also noticed that `Auth.getPreferredMFA` always returns the same MFA method.

So I inspected the source code and noticed that `Auth.getPreferredMFA` **and** `Auth.setPreferredMFA` call the `user.getUserData` method that try to get user data (and preferred MFA method) in the local storage!

https://github.com/aws-amplify/amplify-js/blob/38ea2fc1780135ea6041053be3f5ed8f84edbff8/packages/amazon-cognito-identity-js/src/CognitoUser.js#L1070

So:
1. A user who hasn't configured MFA signs in
-> No preferred MFA method is stored in the local storage
-> `Auth.getPreferredMFA` returns `NOMFA`

2. He sets his MFA method to `SMS` or `TOTP`
-> MFA method is successfully updated but the local storage isn't updated
-> `Auth.getPreferredMFA` returns `NOMFA`

3. He **immediately** sets his MFA method to `NOMFA`
-> `Auth.setPreferredMFA(user, "SMS")` try to get the current preferred MFA method in the local storage but local storage is empty
-> `Auth.getPreferredMFA` returns `NOMFA`

4. MFA method is always set to `SMS` (or `TOTP`) and we get the error `No change for mfa type`.

## A solution to avoid this behavior
We can use the option `bypassCache` of the `user.getUserData` method to bypass local storage.

If you're ok with this, I can make a PR 😄 

---

Could be related to https://github.com/aws-amplify/amplify-js/issues/2519 😉

### PR DESCRIPTION
**Pull request by Gautierrr**
*Issue #, if available:*
fixes #2627 

*Description of changes:*
Add a new parameter to `Auth.setPreferredMFA` and `Auth.getPreferredMFA` methods to allow bypassing cache.

```js
// Get the current preferred MFA method without cache (bypass local storage)
Auth.getPreferredMFA(user, { bypassCache: true });

// Set the preferred MFA method without cache (bypass local storage)
Auth.setPreferredMFA(user, "SMS", { bypassCache: true });
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by powerful23** [comment]
@Gautierrr Hi, sorry about late response. Thanks for your contribution but I think when setting the preferred mfa, the library should always bypass the cache and then update it with the latest result. What do you think?

**Comment by Gautierrr** [comment]
Thanks @powerful23

Yep sounds good. I didn't want to change the default behaviour, just have the choice to override it if needed.

So I agree that the library should always bypass the cache  by default on `Auth.setPreferredMFA` as you said but do you think it would be useful to let the user set `bypassCache` to false anyway to get the old behaviour?

Via the new parameter:

```js
Auth.setPreferredMFA(user, "SMS", { bypassCache: false });
```

**Comment by powerful23** [comment]
@Gautierrr I would say that for `setPreferredMFA` method it should always bypass the cache otherwise will cause the bug. For `getPreferredMFA`method we can add that option to allow the user to choose whether to bypass it or not.

**Comment by Gautierrr** [comment]
Changes made 😉

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -517,25 +517,29 @@ export default class AuthClass {
      * get preferred mfa method
      * @param {CognitoUser} user - the current cognito user
      */
-    public getPreferredMFA(user: CognitoUser | any): Promise<string> {
+    public getPreferredMFA(user: CognitoUser | any, params?: CurrentUserOpts): Promise<string> {
```
Comment (powerful23) [inline packages/auth/src/Auth.ts]: Can you change the params type? It should be something like `GetPreferredMFAOpts`

**Comment by Gautierrr** [comment]
Why update the cached data after successfully signing in? The cache is updated after a signout/signin action.

Do you mean after successfully setting MFA method? Something like this:

```js
public async setPreferredMFA(user, mfaMethod) {
  ...

  const that = this; // line 661

  try {
    await this.cleanCachedItems();
  } catch (e) {
    logger.debug('failed to clear cached items');
  }

  return new Promise<string>((res, rej) => {

  ...

}
```

**Comment by powerful23** [comment]
@Gautierrr because when users calling `currentAuthenticatedUser` after setting mfa successfully, the cached `PreferredMFA` attribute is out-dated. For example, if a user set the MFA from `NOMFA` to `SMS`. After that if he calls `const user = await Auth.currentAuthenticatedUser()`, the `PreferredMFA` attribute will still be `NOMFA`.

The way to update that cached user object is to call `await Auth.currentAuthenticatedUser({bypassCache: true})`. What do you think?

Besides that, can you add some unit tests for the changes? Thanks!

**Comment by Gautierrr** [comment]
Do you mean add `await Auth.currentAuthenticatedUser({bypassCache: true})` after line 661?

I think it's strange. I'd prefer use a dedicated method like in my example. Something like `cleanCachedItems` because it's more explicit we don't want to get the current authenticated user.

**Comment by powerful23** [comment]
@Gautierrr yeah I agree we should not call `currentAuthenticatedUser()`. Instead we only need to call `user.getUserData()` to let the library get the latest user data from the server and this method will also update the cache in the meantime. You can do something similar to https://github.com/aws-amplify/amplify-js/blob/master/packages/auth/src/Auth.ts#L1008-L1010 or even we can extract this part and make it into a private function for now.

**Comment by powerful23** [comment]
@Gautierrr Hi I pushed some change to make sure the latest user data is fetched after setting user mfa preferences successfully. This will make sure the cached data is sync with the server data. Can you take a look and if everything looks fine to you, we will merge this pr.

**Comment by Gautierrr** [comment]
@powerful23 Hi, yep, sounds good to me 👍

**Pull request merged**